pkg/controller/statefulset/stateful_set_control_test.go

```diff
@@ -607,6 +611,255 @@ func TestStatefulSetControl_getSetRevisions(t *testing.T) {
 	}
 }
 
+func TestStatefulSetControlRollingUpdateWithMaxUnavailableInOrderedMode(t *testing.T) {
```

Comment:
Personally I think these tests test too much of the system, which makes them very hard to write and read.

I would like to see them test all (at least probabalistically) orderings of which pods are unavailable.

If I were doing this, I'd probably factor the controller logic to be more testable.

Reply:
Can you transform this test into a test table?

Ideally, you should have an input state, then you just call `UpdateStatefulSet` once, and check the output state.

Reply:
This is still not resolved

Reply:
This is going to take time, since the setup for each UpdateStatefulSet will require some tricks i will need to learn. I think the test coverage is there, will try to get it in post merge

Reply:
I can live with that for now, but it has to be solved before promotion to beta

Reply:
Maybe this can be reconsidered to be able to increase coverage.